commit 47e2bca99e7de52fdaf7e0857e53f81d74a10184 (refs/changes/45/45/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-07-08T14:35:07-07:00 (3 years, 3 months ago)
    
    PUBAPI-1312 Unable to create VM instance with underscore in name

diff --git a/lib/machines.js b/lib/machines.js
index 41c0f34..6a81030 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -1074,6 +1074,13 @@ function loadSSHKeys(req, res, next) {
     });
 }
 
+/*
+ * This should reflect the same policy CNS uses. It's pretty much an aspect of
+ * the public API now, so it can't really be changed.
+ */
+function dnsify(str) {
+    return (str.toLowerCase().replace(/[^a-z0-9-]+/g, '-'));
+}
 
 function create(req, res, next) {
     assert.ok(req.sdc);
@@ -1262,8 +1269,13 @@ function create(req, res, next) {
                 });
                 if (sufs.length > 0) {
                     opts.dns_domain = sufs[0];
-                    /* We've already enforced that the alias is dns-safe. */
-                    opts.hostname = opts.alias;
+                    /*
+                     * Aliases are allowed to have a few characters (eg '_')
+                     * that aren't DNS safe (even though we've enforced that
+                     * they're the right length since the user has CNS turned
+                     * on). So strip them out here the same way CNS will.
+                     */
+                    opts.hostname = dnsify(opts.alias);
                 }
                 cb();
             });
